beacon-chain/core/validators/validator.go

```diff
@@ -310,105 +316,70 @@ func SlashValidator(state *pb.BeaconState, idx uint64) (*pb.BeaconState, error)
-
-	// The maximum balance churn in Gwei (for deposits and exits separately).
-	maxBalChurn := maxBalanceChurn(totalBalance)
-
-	var balChurn uint64
-	var err error
 	vStore.Lock()
 	defer vStore.Unlock()
 	for idx, validator := range state.ValidatorRegistry {
 		// Activate validators within the allowable balance churn.
-		if validator.ActivationEpoch == params.BeaconConfig().FarFutureEpoch &&
-			state.Balances[idx] >= params.BeaconConfig().MaxDepositAmount &&
-			!helpers.IsActiveValidator(validator, currentEpoch) {
-			balChurn += helpers.EffectiveBalance(state, uint64(idx))
-			log.WithFields(logrus.Fields{
-				"index":               idx,
-				"currentBalanceChurn": balChurn,
-				"maxBalanceChurn":     maxBalChurn,
-				"currentEpoch":        currentEpoch - params.BeaconConfig().GenesisEpoch,
-			}).Info("Attempting to activate validator")
-
-			if balChurn > maxBalChurn {
-				break
-			}
-			state, err = ActivateValidator(state, uint64(idx), false)
-			if err != nil {
-				return nil, fmt.Errorf("could not activate validator %d: %v", idx, err)
-			}
-			vStore.activatedValidators[updatedEpoch] =
-				append(vStore.activatedValidators[updatedEpoch], uint64(idx))
+		if validator.ActivationEligibilityEpoch == params.BeaconConfig().FarFutureEpoch &&
+			validator.EffectiveBalance >= params.BeaconConfig().MaxEffectiveBalance {
+			validator.ActivationEligibilityEpoch = currentEpoch
+		}
+		if helpers.IsActiveValidator(validator, currentEpoch) &&
+			validator.EffectiveBalance <= params.BeaconConfig().EjectionBalance {
+			state = ExitValidator(state, uint64(idx))
 		}
 	}
+	// queue validators eligible for activation and not dequeued for activation prior to finalized epoch
 
-	balChurn = 0
+	activationQueue := []queueElement{}
 	for idx, validator := range state.ValidatorRegistry {
-		// Exit validators within the allowable balance churn.
-		if validator.ExitEpoch == params.BeaconConfig().FarFutureEpoch &&
-			validator.StatusFlags == pb.Validator_INITIATED_EXIT {
-			balChurn += helpers.EffectiveBalance(state, uint64(idx))
-			if balChurn > maxBalChurn {
-				break
-			}
-			state = ExitValidator(state, uint64(idx))
-			vStore.exitedValidators[updatedEpoch] =
-				append(vStore.exitedValidators[updatedEpoch], uint64(idx))
+		if validator.ActivationEligibilityEpoch != params.BeaconConfig().FarFutureEpoch &&
+			validator.ActivationEpoch >= helpers.DelayedActivationExitEpoch(state.FinalizedEpoch) {
+			qe := queueElement{idx: idx,
+				aEE: validator.ActivationEligibilityEpoch}
+			activationQueue = append(activationQueue, qe)
 		}
 	}
-	state.ValidatorRegistryUpdateEpoch = currentEpoch
-	return state, nil
+	sort.Slice(activationQueue, func(i, j int) bool { return activationQueue[i].aEE < activationQueue[j].aEE })
+	// dequeued validators for activation up to churn limit (without resetting activation epoch)
+	limit := uint64(len(activationQueue))
+	if helpers.ChurnLimit(state) < limit {
+		limit = helpers.ChurnLimit(state)
```

Comment:
This line is not tested, can we add test to cover this line?

Reply:
added tests